codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -284,194 +316,542 @@ private void writeFromStreamChannel0(Object msg, boolean flush, ChannelPromise p
+                }
+                for (;;) {
+                    final Object msg = in.current();
+                    if (msg == null) {
+                        break;
+                    }
+                    if (msg == REEVALUATE_WRITABILITY_MESSAGE) {
+                        in.remove();
+                        continue;
+                    }
+
+                    final int bytes = sizeEstimatorHandle.size(msg);
+
+                    /*
+                     * The flow control window needs to be decrement before stealing the message from the buffer (and
+                     * thereby decrementing the number of pending bytes). Else, when calling steal() the number of
+                     * pending bytes could  be less than the writebuffer watermark (=flow control window) and thus
+                     * trigger a  writability change.
+                     *
+                     * This code must never trigger a writability change. Only reading window updates or channel writes
+                      * may change the channel's writability.
+                     */
+                    incrementOutboundFlowControlWindow(-bytes);
+
+                    final ChannelPromise promise = in.steal();
+
+                    if (bytes > 0) {
+                        promise.addListener(new ReturnFlowControlWindowOnFailureListener(bytes));
+                    }
+
+                    if (promise.isCancelled()) {
+                        // Write was cancelled so skip it.
+                        continue;
+                    }
+
+                    // TODO(buchgr): Should we also the change the writability if END_STREAM is set?
+                    try {
+                        doWrite(msg, promise);
+                    } catch (Throwable t) {
+                        promise.tryFailure(t);
+                    }
+                }
+            } finally {
+                // ensure we always flush even if the write loop throws.
+                mayFlush();
+            }
+        }
+
+        private void mayFlush() {
+            // If we are current channelReadComplete(...) call we should just mark this Channel with a flush pending.
+            // We will ensure we trigger ctx.flush() after we processed all Channels later on and so aggregate the
+            // flushes. This is done as ctx.flush() is expensive when as it may trigger an write(...) or writev(...)
+            // operation on the socket.
+            if (inFireChannelReadComplete) {
+                flushPending = true;
+            } else {
+                ctx.flush();
+            }
+        }
+
+        /**
+         * Receive a read message. This does not notify handlers unless a read is in progress on the
+         * channel.
+         *
+         * If returns {@code true} we will also have a flush pending for this channel.
+         */
+        boolean fireChildRead(final Object msg) {
+            assert eventLoop().inEventLoop();
+
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Merge this if statement with the enclosing one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1066)